File: android-sample-java/app/src/main/java/io/logto/demo4j/viewmodel/LogtoViewModel.java
package io.logto.demo4j.viewmodel;

import android.app.Activity;
import android.app.Application;
import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import io.logto.sdk.android.LogtoClient;
import io.logto.sdk.android.exception.LogtoException;
import io.logto.sdk.android.type.AccessToken;
import io.logto.sdk.android.type.LogtoConfig;
import io.logto.sdk.core.type.IdTokenClaims;

public class LogtoViewModel extends AndroidViewModel {

    public LogtoViewModel(@NonNull Application application) {
        super(application);
    }

    private LogtoConfig logtoConfig = new LogtoConfig(
            "https://logto.dev",
            "94fKrpteyMI6BAy9K3pdX",
            null,
            null,
            true
    );

    private final LogtoClient logtoClient = new LogtoClient(logtoConfig, getApplication());

    private final MutableLiveData<Boolean> _authenticated = new MutableLiveData<>(logtoClient.isAuthenticated());
    public LiveData<Boolean> isAuthenticated() {
        return _authenticated;
    }

    private final MutableLiveData<AccessToken> _accessToken = new MutableLiveData<>();
    public LiveData<AccessToken> getAccessToken() {
        return _accessToken;
    }

    private final MutableLiveData<IdTokenClaims> _idTokenClaims = new MutableLiveData<>();
    public LiveData<IdTokenClaims> getIdTokenClaims() {
        return _idTokenClaims;
    }

    private final MutableLiveData<LogtoException> _logtoException = new MutableLiveData<>();
    public LiveData<LogtoException> getLogtoException() {
        return _logtoException;
    }


    public void signIn(Activity context) {
        logtoClient.signIn(
                context,
                "io.logto.android://io.logto.sample/callback",
                logtoException -> {
                    if (logtoException != null) {
                        _logtoException.postValue(logtoException);
                        return;
                    }
                    _authenticated.postValue(logtoClient.isAuthenticated());
                });
    }

    public void signOut() {
        logtoClient.signOut(logtoException -> {
            if (logtoException != null) {
                _logtoException.postValue(logtoException);
            }
            _authenticated.postValue(logtoClient.isAuthenticated());
        });
    }

    public void retrieveAccessToken() {
        logtoClient.getAccessToken((logtoException, accessToken) -> {
            if (logtoException != null) {
                _logtoException.postValue(logtoException);
                return;
            }
            _accessToken.postValue(accessToken);
        });
    }

    public void retrieveIdTokenClaims() {
        logtoClient.getIdTokenClaims((logtoException, idTokenClaims) -> {
            if (logtoException != null) {
                _logtoException.postValue(logtoException);
                return;
            }
            _idTokenClaims.postValue(idTokenClaims);
        });
    }

    public void clearException() {
        _logtoException.postValue(null);
    }
}

File: android-sample-kotlin/app/src/main/kotlin/io/logto/demo/viewmodel/LogtoViewModel.kt
package io.logto.demo.viewmodel

import android.app.Activity
import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import io.logto.sdk.android.LogtoClient
import io.logto.sdk.android.exception.LogtoException
import io.logto.sdk.android.type.AccessToken
import io.logto.sdk.android.type.LogtoConfig
import io.logto.sdk.core.type.IdTokenClaims

class LogtoViewModel(application: Application) : AndroidViewModel(application) {

    private val logtoConfig = LogtoConfig(
        endpoint = "https://logto.dev",
        clientId = "94fKrpteyMI6BAy9K3pdX",
        scopes = null,
        resources = null,
        usingPersistStorage = true
    )

    private val logtoClient = LogtoClient(logtoConfig, getApplication())

    private val _authenticated = MutableLiveData(logtoClient.isAuthenticated)
    val authenticated: LiveData<Boolean>
        get() = _authenticated

    private val _accessToken = MutableLiveData<AccessToken>()
    val accessToken: LiveData<AccessToken>
        get() = _accessToken

    private val _idTokenClaims = MutableLiveData<IdTokenClaims>()
    val idTokenClaims: LiveData<IdTokenClaims>
        get() = _idTokenClaims

    private val _logtoException = MutableLiveData<LogtoException>()
    val logtoException: LiveData<LogtoException>
        get() = _logtoException

    fun signInWithBrowser(context: Activity) {
        logtoClient.signInWithBrowser(context, "io.logto.android://io.logto.sample/callback",) {
            it?.let { _logtoException.postValue(it) } ?: _authenticated.postValue(logtoClient.isAuthenticated)
        }
    }

    fun signOut() {
        logtoClient.signOut {
            it?.let { _logtoException.postValue(it) }
            _authenticated.postValue(logtoClient.isAuthenticated)
        }
    }

    fun getAccessToken() {
        logtoClient.getAccessToken { logtoException, accessToken ->
            logtoException?.let { _logtoException.postValue(it) } ?: _accessToken.postValue(accessToken)
        }
    }

    fun getIdTokenClaims() {
        logtoClient.getIdTokenClaims { logtoException, idTokenClaims ->
            logtoException?.let { _logtoException.postValue(it) } ?: _idTokenClaims.postValue(idTokenClaims)
        }
    }

    fun clearException() {
        _logtoException.postValue(null)
    }
}
